Accept application/jwk-set+json as content type. [SDK-2131]

# Description

The JWKS endpoint I am using supplies the key set with the content type `application/jwk-set+json`, as is in line with the [spec](https://tools.ietf.org/html/rfc7517). I saw that since #59 the header is set to accept only the `application/json`-type though. This leads to the call to `UrlJwkProvider.getAll()` throwing a `SigningKeyNotFoundException`, because `406: Not acceptable` is returned by the server, as the content type in the accept-header is not matching the content type of the key set. I would recommend adding the possibility to supply a content type for the accept-header to the `UrlJwkProvider` or allowing it to accept either `application/json` or `application/jwk-set+json`

### Reproduction

To reproduce this simply try using the library with a JWKS-Endpoint that supplies the key set with the content type `application/jwk-set+json`.

Stacktrace (JWK URL and android package are modified):
```
com.auth0.jwk.SigningKeyNotFoundException: Cannot obtain jwks from url https://example.com/app/authorization/v1/jwks
        at com.auth0.jwk.UrlJwkProvider.getJwks(UrlJwkProvider.java:112)
        at com.auth0.jwk.UrlJwkProvider.getAll(UrlJwkProvider.java:118)
        at com.auth0.jwk.UrlJwkProvider.get(UrlJwkProvider.java:136)
        at com.myapp.fragments.setup.OAuthFragment.validateToken(OAuthFragment.kt:167)
        at com.myapp.fragments.setup.OAuthFragment.access$validateToken(OAuthFragment.kt:53)
        at com.myapp.fragments.setup.OAuthFragment$onActivityResult$1$1$1.invokeSuspend(OAuthFragment.kt:110)
        at com.myapp.fragments.setup.OAuthFragment$onActivityResult$1$1$1.invoke(Unknown Source:10)
        at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:155)
        at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1)
        at com.myapp.fragments.setup.OAuthFragment$onActivityResult$1$1.invokeSuspend(OAuthFragment.kt:106)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
     Caused by: java.io.FileNotFoundException: https://example.com/app/authorization/v1/jwks
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:255)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getInputStream(DelegatingHttpsURLConnection.java:211)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:30)
        at com.auth0.jwk.UrlJwkProvider.getJwks(UrlJwkProvider.java:108)
        at com.auth0.jwk.UrlJwkProvider.getAll(UrlJwkProvider.java:118) 
        at com.auth0.jwk.UrlJwkProvider.get(UrlJwkProvider.java:136) 
        at com.myapp.fragments.setup.OAuthFragment.validateToken(OAuthFragment.kt:167) 
        at com.myapp.fragments.setup.OAuthFragment.access$validateToken(OAuthFragment.kt:53) 
        at com.myapp.fragments.setup.OAuthFragment$onActivityResult$1$1$1.invokeSuspend(OAuthFragment.kt:110) 
        at com.myapp.fragments.setup.OAuthFragment$onActivityResult$1$1$1.invoke(Unknown Source:10) 
        at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91) 
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:155) 
        at kotlinx.coroutines.BuildersKt.withContext(Unknown Source:1) 
        at com.myapp.fragments.setup.OAuthFragment$onActivityResult$1$1.invokeSuspend(OAuthFragment.kt:106) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742) 
```

### Environment

- **Version of this library used:** 0.9.0
- I am using the library in Android.

--- a/src/main/java/com/auth0/jwk/JwkProviderBuilder.java
+++ b/src/main/java/com/auth0/jwk/JwkProviderBuilder.java
@@ -2,6 +2,7 @@ package com.auth0.jwk;
 
 import java.net.Proxy;
 import java.net.URL;
+import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
 import static com.auth0.jwk.UrlJwkProvider.urlForDomain;
@@ -20,6 +21,7 @@ public class JwkProviderBuilder {
     private boolean cached;
     private BucketImpl bucket;
     private boolean rateLimited;
+    private Map<String, String> headers;
 
     /**
      * Creates a new Builder with the given URL where to load the jwks from.
@@ -127,12 +129,23 @@ public class JwkProviderBuilder {
     }
 
     /**
+     * Sets the headers to send on the request. Any headers set here will override the default headers ("accept" -> "application/json")
+     *
+     * @param headers a map of header keys to values to send on the request.
+     * @return this builder instance
+     */
+    public JwkProviderBuilder headers(Map<String, String> headers) {
+        this.headers = headers;
+        return this;
+    }
+
+    /**
      * Creates a {@link JwkProvider}
      *
      * @return a newly created {@link JwkProvider}
      */
     public JwkProvider build() {
-        JwkProvider urlProvider = new UrlJwkProvider(url, null, null, proxy);
+        JwkProvider urlProvider = new UrlJwkProvider(url, null, null, proxy, headers);
         if (this.rateLimited) {
             urlProvider = new RateLimitedJwkProvider(urlProvider, bucket);
         }

--- a/src/main/java/com/auth0/jwk/UrlJwkProvider.java
+++ b/src/main/java/com/auth0/jwk/UrlJwkProvider.java
@@ -4,6 +4,7 @@ import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.ObjectReader;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.collect.Lists;
+import com.google.common.io.ByteStreams;
 
 import java.io.IOException;
 import java.io.InputStream;
@@ -13,6 +14,8 @@ import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.net.URLConnection;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
@@ -30,6 +33,7 @@ public class UrlJwkProvider implements JwkProvider {
 
     final URL url;
     final Proxy proxy;
+    final Map<String, String> headers;
     private final Integer connectTimeout;
     private final Integer readTimeout;
 
@@ -40,7 +44,7 @@ public class UrlJwkProvider implements JwkProvider {
      * @param url to load the jwks
      */
     public UrlJwkProvider(URL url) {
-        this(url, null, null, null);
+        this(url, null, null, null, null);
     }
 
     /**
@@ -52,6 +56,19 @@ public class UrlJwkProvider implements JwkProvider {
      * @param proxy          proxy server to use when making the connection
      */
     public UrlJwkProvider(URL url, Integer connectTimeout, Integer readTimeout, Proxy proxy) {
+        this(url, connectTimeout, readTimeout, proxy, null);
+    }
+
+    /**
+     * Creates a provider that loads from the given URL using custom request headers.
+     *
+     * @param url            to load the jwks
+     * @param connectTimeout connection timeout in milliseconds (default is null)
+     * @param readTimeout    read timeout in milliseconds (default is null)
+     * @param proxy          proxy server to use when making the connection (default is null)
+     * @param headers        a map of request header keys to values to send on the request. Default is "accept -> application/json".
+     */
+    public UrlJwkProvider(URL url, Integer connectTimeout, Integer readTimeout, Proxy proxy, Map<String, String> headers) {
         checkArgument(url != null, "A non-null url is required");
         checkArgument(connectTimeout == null || connectTimeout >= 0, "Invalid connect timeout value '" + connectTimeout + "'. Must be a non-negative integer.");
         checkArgument(readTimeout == null || readTimeout >= 0, "Invalid read timeout value '" + readTimeout + "'. Must be a non-negative integer.");
@@ -60,8 +77,10 @@ public class UrlJwkProvider implements JwkProvider {
         this.proxy = proxy;
         this.connectTimeout = connectTimeout;
         this.readTimeout = readTimeout;
-        
         this.reader = new ObjectMapper().readerFor(Map.class);
+
+        this.headers = (headers == null) ?
+            Collections.singletonMap("Accept", "application/json") : headers;
     }
 
     /**
@@ -72,7 +91,7 @@ public class UrlJwkProvider implements JwkProvider {
      * @param readTimeout    read timeout in milliseconds (null for default)
      */
     public UrlJwkProvider(URL url, Integer connectTimeout, Integer readTimeout) {
-        this(url, connectTimeout, readTimeout, null);
+        this(url, connectTimeout, readTimeout, null, null);
     }
 
     /**
@@ -118,8 +137,11 @@ public class UrlJwkProvider implements JwkProvider {
             if (readTimeout != null) {
                 c.setReadTimeout(readTimeout);
             }
-            c.setRequestProperty("Accept", "application/json");
-            
+
+            for (Map.Entry<String, String> entry : headers.entrySet()) {
+                c.setRequestProperty(entry.getKey(), entry.getValue());
+            }
+
             try (InputStream inputStream = c.getInputStream()) {
                 return reader.readValue(inputStream);
             }

--- a/src/test/java/com/auth0/jwk/JwkProviderBuilderTest.java
+++ b/src/test/java/com/auth0/jwk/JwkProviderBuilderTest.java
@@ -7,6 +7,8 @@ import org.junit.rules.ExpectedException;
 import java.net.InetSocketAddress;
 import java.net.Proxy;
 import java.net.URL;
+import java.util.Collections;
+import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
 import static com.auth0.jwk.UrlJwkProvider.WELL_KNOWN_JWKS_PATH;
@@ -157,5 +159,19 @@ public class JwkProviderBuilderTest {
         assertThat(provider, notNullValue());
         UrlJwkProvider urlJwkProvider = (UrlJwkProvider) provider;
         assertThat(urlJwkProvider.proxy, equalTo(proxy));
+    }
+
+    @Test
+    public void shouldCreateForUrlWithCustomHeaders() throws Exception {
+        URL url = new URL(normalizedDomain + WELL_KNOWN_JWKS_PATH);
+        Map<String, String> headers = Collections.singletonMap("header", "value");
+        JwkProvider provider = new JwkProviderBuilder(url)
+                .rateLimited(false)
+                .cached(false)
+                .headers(headers)
+                .build();
+        assertThat(provider, notNullValue());
+        UrlJwkProvider urlJwkProvider = (UrlJwkProvider) provider;
+        assertThat(urlJwkProvider.headers, equalTo(headers));
     }
 }

--- a/src/test/java/com/auth0/jwk/UrlJwkProviderTest.java
+++ b/src/test/java/com/auth0/jwk/UrlJwkProviderTest.java
@@ -11,6 +11,8 @@ import org.mockito.stubbing.Answer;
 import java.io.IOException;
 import java.lang.ref.WeakReference;
 import java.net.*;
+import java.util.Collections;
+import java.util.List;
 
 import static com.auth0.jwk.UrlJwkProvider.WELL_KNOWN_JWKS_PATH;
 import static org.hamcrest.Matchers.*;
@@ -275,6 +277,14 @@ public class UrlJwkProviderTest {
         assertThat(mockFactory.urlUsed.get(), is(url));
         assertThat(mockFactory.proxyUsed.get(), is(nullValue()));
 
+        // Test creation: custom headers
+        UrlJwkProvider urlJwkProviderWithHeaders = new UrlJwkProvider(url, connectTimeout, readTimeout, null,
+            Collections.singletonMap("Accept", "application/jwks-set+json"));
+        Jwk hJwk = urlJwkProviderWithHeaders.get("NkJCQzIyQzRBMEU4NjhGNUU4MzU4RkY0M0ZDQzkwOUQ0Q0VGNUMwQg");
+        assertNotNull(hJwk);
+        assertThat(mockFactory.urlUsed.get(), is(url));
+        assertThat(mockFactory.proxyUsed.get(), is(nullValue()));
+
         //Test creation: with Proxy
         Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress("localhost", 8080));
         URL pUrl = new URL("mock://localhost");
@@ -291,15 +301,16 @@ public class UrlJwkProviderTest {
         //Test 1: Configuration
         //Request Timeout assertions
         ArgumentCaptor<Integer> connectTimeoutCaptor = ArgumentCaptor.forClass(Integer.class);
-        verify(urlConnection, times(2)).setConnectTimeout(connectTimeoutCaptor.capture());
+        verify(urlConnection, times(3)).setConnectTimeout(connectTimeoutCaptor.capture());
         assertThat(connectTimeoutCaptor.getValue(), is(connectTimeout));
 
         ArgumentCaptor<Integer> readTimeoutCaptor = ArgumentCaptor.forClass(Integer.class);
-        verify(urlConnection, times(2)).setReadTimeout(readTimeoutCaptor.capture());
+        verify(urlConnection, times(3)).setReadTimeout(readTimeoutCaptor.capture());
         assertThat(readTimeoutCaptor.getValue(), is(readTimeout));
 
         //Request Headers assertions
         verify(urlConnection, times(2)).setRequestProperty("Accept", "application/json");
+        verify(urlConnection, times(1)).setRequestProperty("Accept", "application/jwks-set+json");
 
         //Test 2: Network errors
         Exception capturedException = null;

